### PR TITLE
Catch up to spectator-cpp

### PR DESCRIPTION
**Pull request by fvallenilla**
## Description
This change updates the fork to catch up with Netflix/spectator-cpp. 

The main change that was needed was the commit that replaces invalid characters in ids, but I pulled in another couple of changes that happened before that. That fix is needed to address an issue where some IPC metrics are being silently dropped due to having invalid characters in tags.

NOTE: There are a couple departures from the upstream repo:
1. I kept Bazel as the build system to facilitate import into proxyd
2. I checked `spectator/valid_chars.inc` into git, rather than having it be auto-generated. This was also to facilitate import into proxyd.

## How was this tested?
1. Verified that spectator-cpp continues to build with Bazel
4. Ran the unit tests
5. Built proxyd
6. Ran custom build of proxyd and verified that metrics with invalid tags/values were sanitized and made their way to Atlas.

The repo's GitHub actions don't run against forks, so I built a container to build and test the library based on the CI script:

```Dockerfile
FROM ubuntu:22.04

# Set environment variables
ENV CC=gcc-11
ENV CXX=g++-11
ENV LANG=en_US.UTF-8

# Install system dependencies
RUN apt-get update && \
    apt-get install -y \
    curl \
    gnupg \
    software-properties-common \
    binutils-dev \
    g++-11 \
    git \
    wget \
    && rm -rf /var/lib/apt/lists/*

# Install Bazel
RUN wget -O /usr/local/bin/bazel https://github.com/bazelbuild/bazelisk/releases/latest/download/bazelisk-linux-$([ $(uname -m) = "aarch64" ] && echo "arm64" || echo "amd64")
RUN chmod +x /usr/local/bin/bazel

# Create a directory for the project
WORKDIR /app
COPY . .

# Set the output user root for Bazel cache
ENV BAZEL_OUTPUT_USER_ROOT=/root/.cache/bazel

# Build the project
RUN bazel --batch build --jobs=2 --config asan spectator_test spectator

# Run the tests
CMD ["bash", "-c", "GTEST_COLOR=1 ./bazel-bin/spectator_test"]
```

### CODE REVIEW

**Comment by cancecen** [comment]
The write up is excellent, thanks for all the details.

File: README.md [hunk 1]
```diff
@@ -2,13 +2,8 @@
...
 # Spectator-cpp
 
-Simple library for instructing code to record dimensional time series.  It sends all activity to
-a [spectatord](https://github.com/Netflix-Skunkworks/spectatord) sidecar.
-
-## Description
-
-This implements a basic [Spectator](https://github.com/Netflix/spectator) library for instrumenting
-C++ applications, sending all metric activity to a sidecar. 
+This implements a basic [Spectator](https://github.com/Netflix/spectator) library for instrumenting Go applications. It
```
Comment (cancecen) [inline README.md]: Not sure if this is accurate, but I guess it comes from upstream?

File: tools/gen_valid_chars.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// generate the atlas valid charsets
+
+#include <array>
+#include <fstream>
+
+void dump_array(std::ostream& os, const std::string& name, const std::array<bool, 256>& chars) {
```
Comment (cancecen) [inline tools/gen_valid_chars.cc:6]: how is this file useful?

**Review by cancecen — COMMENTED** [review]
Do you want to also check in the docker run script you wrote for the tests?

File: README.md (hunk 1 above)
Comment (fvallenilla) [reply README.md]: Good eye, that did come from there:
https://github.com/Netflix/spectator-cpp/blob/8b9d4098a2b520660bcab56e482b217d78c99c41/README.md?plain=1#L5

File: tools/gen_valid_chars.cc (hunk 2 above)
Comment (fvallenilla) [reply tools/gen_valid_chars.cc:6]: That file is used to generate `spectator/valid_chars.inc`, which is not checked in by the upstream. 

The upstream library calls this tool at build time to generate `valid_chars.inc`. I ended up calling it locally to generate the file.

File: README.md (hunk 1 above)
Comment (fvallenilla) [reply README.md]: I edited it back to C++ in this PR

**Pull request merged**